File: prthgnode/src/common/classes/DateAndTime.ts
/**
 * @file DateAndTime.ts
 * @author Prana Ron (prthg@proton.me)
 * 
 * @copyright Copyright (c) 2023
 * 
 * @contributors:
 * - Prana Ron @prothegee (prthg@proton.me)
 * 
*/
import CEndUser from "./EndUser";
import { EDateAndTimeFormat } from "../enums/DateEnums";


/**
 * @brief date and time class
 * 
 * @note extend from CEndUser
*/
export default class CDateAndTime extends CEndUser
{
    constructor()
    {
        super();
    }


    /**
     * @brief get current time
     * 
     * @note with at least 2 character
     * 
     * @param dateAndTimeFormat
     * 
     * @returns string
    */
    public GetTime(dateAndTimeFormat:EDateAndTimeFormat):string
    {
        let result:string;


        switch (dateAndTimeFormat)
        {
            case EDateAndTimeFormat.FULL_TIMESTAMP:
            {
                const year = this.m_YYYY;
                const yearStr = year.toString();

                const month = this.m_MM;
                const monthStr = (month < 10) ? `0${month.toString()}` : month.toString();

                const day = this.m_DD;
                const dayStr = (day < 10) ? `0${day.toString()}` : day.toString();

                const hour = this.m_hh;
                const hourStr = (hour < 10) ? `0${hour.toString()}` : hour.toString();

                const minute = this.m_mm;
                const minuteStr = (minute < 10) ? `0${minute.toString()}` : minute.toString();

                const second = this.m_ss;
                const secondStr = (second < 10) ? `0${second.toString()}` : second.toString();


                result = `${yearStr}${monthStr}${dayStr}${hourStr}${minuteStr}${secondStr}`;
            }
            break;




            case EDateAndTimeFormat.YYYYMMDD:
            {
                const year = this.m_YYYY;
                const yearStr = year.toString();

                const month = this.m_MM;
                const monthStr = (month < 10) ? `0${month.toString()}` : month.toString();

                const day = this.m_DD;
                const dayStr = (day < 10) ? `0${day.toString()}` : day.toString();


                result = `${yearStr}${monthStr}${dayStr}`;
            }
            break;




            case EDateAndTimeFormat.MM:
            {
                result = (this.m_MM < 10) ? `0${this.m_MM.toString()}` : this.m_MM.toString();
            }
            break;




            case EDateAndTimeFormat.DD:
            {
                result = (this.m_DD < 10) ? `0${this.m_DD.toString()}` : this.m_DD.toString();
            }
            break;




            case EDateAndTimeFormat.hhmmss:
            {
                const hour = this.m_hh;
                const hourStr = (hour < 10) ? `0${hour.toString()}` : hour.toString();

                const minute = this.m_mm;
                const minuteStr = (minute < 10) ? `0${minute.toString()}` : minute.toString();

                const second = this.m_ss;
                const secondStr = (second < 10) ? `0${second.toString()}` : second.toString();


                result = `${hourStr}:${minuteStr}:${secondStr}`;
            }
            break;




            case EDateAndTimeFormat.hh:
            {
                result = (this.m_hh < 10) ? `0${this.m_hh.toString()}` : this.m_hh.toString();
            }
            break;




            case EDateAndTimeFormat.mm:
            {
                result = (this.m_mm < 10) ? `0${this.m_mm.toString()}` : this.m_mm.toString();
            }
            break;




            case EDateAndTimeFormat.ss:
            {
                result = (this.m_ss < 10) ? `0${this.m_ss.toString()}` : this.m_ss.toString();
            }
            break;




            default:
            {
                result = "ERR";
            }
            break;
        }


        return result;
    }
}

File: prthgcpp/src/prthgcpp/classes/DateAndTime.cc
#include "prthgcpp/classes/DateAndTime.h"


prthgcpp::CDateAndTime::CDateAndTime()
{
}

prthgcpp::CDateAndTime::~CDateAndTime()
{
}








std::string prthgcpp::CDateAndTime::GetTime(const prthgcpp::ECDateAndTimeFormat dateAndTimeFormat) const
{
    std::string result;


    switch (dateAndTimeFormat)
    {
        case ECDateAndTimeFormat::UTC_TIME:
        {
            auto now = std::chrono::system_clock::now();
            auto itt = std::chrono::system_clock::to_time_t(now);

            std::stringstream ss;
            ss << std::put_time(gmtime(&itt), "%F %T UTC");

            result = ss.str();
        }
        break;




        case ECDateAndTimeFormat::FULL_TIMESTAMP:
        {
            auto year_str   = [&](){ return std::to_string(m_year); }();
            auto month_str  = [&](){ return std::to_string(m_month); }();
            auto day_str    = [&](){ return std::to_string(m_day); }();
            auto hour_str   = [&](){ return std::to_string(m_hour); }();
            auto minute_str = [&](){ return std::to_string(m_minute); }();
            auto second_str = [&](){ return std::to_string(m_second); }();

            auto year_fmt   = [&](){ return year_str; }();
            auto month_fmt  = [&](){ return (month_str.size() < 2) ? "0" + month_str : month_str; }();
            auto day_fmt    = [&](){ return (month_str.size() < 2) ? "0" + day_str : day_str; }();
            auto hour_fmt   = [&](){ return (hour_str.size() < 2) ? "0" + hour_str : hour_str; }();
            auto minute_fmt = [&](){ return (minute_str.size() < 2) ? "0" + minute_str : minute_str; }();
            auto second_fmt = [&](){ return (second_str.size() < 2) ? "0" + second_str : second_str; }();

            result = year_fmt + month_fmt + day_fmt + hour_fmt + minute_fmt + second_fmt;
        }
        break;




        case ECDateAndTimeFormat::YYYYMMDD:
        {
            auto year_str  = [&](){ return std::to_string(m_year); }();
            auto month_str = [&](){ return std::to_string(m_month); }();
            auto day_str   = [&](){ return std::to_string(m_day); }();

            auto year_fmt  = [&](){ return year_str; }();
            auto month_fmt = [&](){ return (month_str.size() < 2) ? "0" + month_str : month_str; }();
            auto day_fmt   = [&](){ return (month_str.size() < 2) ? "0" + day_str : day_str; }();

            result = year_fmt + month_fmt + day_fmt;
        }
        break;




        case ECDateAndTimeFormat::YYYY:
        {
            result = std::to_string(m_year);
        }
        break;




        case ECDateAndTimeFormat::MM:
        {
            auto month = std::to_string(m_month);

            result = [&](){ return (month.size() < 2) ? "0" + month : month; }();
        }
        break;




        case ECDateAndTimeFormat::DD:
        {
            auto day = std::to_string(m_day);

            result = [&](){ return (day.size() < 2) ? "0" + day : day; }();
        }
        break;




        case ECDateAndTimeFormat::hhmmss:
        {
            auto hour_str   = [&](){ return std::to_string(m_hour); }();
            auto minute_str = [&](){ return std::to_string(m_minute); }();
            auto second_str = [&](){ return std::to_string(m_second); }();

            auto hour_fmt   = [&](){ return (hour_str.size() < 2) ? "0" + hour_str : hour_str; }();
            auto minute_fmt = [&](){ return (minute_str.size() < 2) ? "0" + minute_str : minute_str; }();
            auto second_fmt = [&](){ return (second_str.size() < 2) ? "0" + second_str : second_str; }();

            result = hour_fmt + ":" + minute_fmt + ":" + second_fmt;
        }
        break;




        case ECDateAndTimeFormat::hh:
        {
            auto hour = std::to_string(m_hour);

            result = [&](){ return (hour.size() < 2) ? "0" + hour : hour; }();
        }
        break;




        case ECDateAndTimeFormat::mm:
        {
            auto minute = std::to_string(m_minute);

            result = [&](){ return (minute.size() < 2) ? "0" + minute : minute; }();
        }
        break;




        case ECDateAndTimeFormat::ss:
        {
            auto second = std::to_string(m_second);

            result = [&](){ return (second.size() < 2) ? "0" + second : second; }();
        }
        break;
    }


    return result;
}